### PR TITLE
code review

### PR DESCRIPTION
**Pull request by emunac**

### CODE REVIEW

File: face_detection.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    def __init__(self, box: list, pred: int, idx: int, v_len: int):
+        self.current_center = np.array([(box[0] + box[2])/2, (box[1] + box[3])/2])
+        self.predictions_in_frames = [None] * v_len
+        self.predictions_in_frames[idx] = pred
+        self.coordinates_in_frames = [None] * v_len
+        self.coordinates_in_frames[idx] = box
+
+
+def assign(face_objects_list: list, boxes, predictions, idx, v_len) -> list:
+    max_assign_distance = 50
```
Comment (anatshk) [inline face_detection.py]: it is customary to place constants (a = 5, b = 'hello', etc) either in the top of file or in another `consts.py` file and import from there.

also - const variable names are uppercase 

--> MAX_ASSIGN_DISTANCE = 50

File: face_detection.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    if len(boxes) != len(predictions):
+        print("error, faces and predictions should have same length")
```
Comment (anatshk) [inline face_detection.py]: print error --> raise Exception?

File: face_detection.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        return face_objects_list
+
+    for box, prediction in zip(boxes, predictions):
+        center = np.array([(box[0] + box[2])/2, (box[1] + box[3])/2])
```
Comment (anatshk) [inline face_detection.py]: code duplication - functio please 
`np.array([(box[0] + box[2])/2, (box[1] + box[3])/2])`

File: face_detection.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+        min_distance = max_assign_distance
+        closest_face = None
```
Comment (anatshk) [inline face_detection.py]: min_distance, closest_face = max_assign_distance, None

File: face_detection.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            face = FaceObject(box, prediction, idx, v_len)
+            face_objects_list.append(face)
+
+    return face_objects_list
+
+
+def correct(face_objects_list: list):
+    min_frame_appearance = 10
+    window_size = 10
+    sensitivity = 0
```
Comment (anatshk) [inline face_detection.py]: consts?

File: face_detection.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    for face in face_objects_list:
+        count_frame_appearance = sum(x is not None for x in face.predictions_in_frames)
+        if count_frame_appearance < min_frame_appearance:
+            face_objects_list.remove(face)
+            continue
+
+        for i, x in enumerate(face.predictions_in_frames):
+            stri = str(x.item()) if x is not None else ''
+            print('frame #', i, 'frame detect:', stri)
+            # print(x.item(), i)
```
Comment (anatshk) [inline face_detection.py]: commented out code - delete please

File: face_detection.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+        corrected_predictions: list = face.predictions_in_frames
+        half_window_size = int(window_size/2)
+        for i in range(half_window_size, len(face.predictions_in_frames) - half_window_size):
+            if sum(x is None for x in
```
Comment (anatshk) [inline face_detection.py]: `sum(x is None for x in` seems like a duplication - can it be a small function with the end of the array as input?

File: face_detection.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+                   face.predictions_in_frames[i - half_window_size: i + 1 + half_window_size]) > 0:
+                continue
+
+            majority = sum(face.predictions_in_frames[i - half_window_size: i + 1 + half_window_size])
+            if majority < half_window_size - sensitivity:
+                corrected_predictions[i] = 0
+            else:
+                corrected_predictions[i] = 1
```
Comment (anatshk) [inline face_detection.py]: one line if please

File: face_detection.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+        face.predictions_in_frames = corrected_predictions
+
+
+image_size = 64
+mtcnn = MTCNN(image_size=image_size, margin=0, post_process=False)
+
+root = Tk()
+dir_path = '/home/emuna/PycharmProjects/SheCodes/examples/'
```
Comment (anatshk) [inline face_detection.py]: maybe the path should not be hardcoded, but an input / cli parameter

File: face_detection.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+            majority = sum(face.predictions_in_frames[i - half_window_size: i + 1 + half_window_size])
+            if majority < half_window_size - sensitivity:
+                corrected_predictions[i] = 0
+            else:
+                corrected_predictions[i] = 1
+
+        face.predictions_in_frames = corrected_predictions
+
+
+image_size = 64
```
Comment (anatshk) [inline face_detection.py]: const

File: face_detection.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+mtcnn = MTCNN(image_size=image_size, margin=0, post_process=False)
+
+root = Tk()
+dir_path = '/home/emuna/PycharmProjects/SheCodes/examples/'
+filepath = filedialog.askopenfilename(initialdir=dir_path, title="Select file",
+                                      filetypes=(("video files", "*.mp4"), ("all files", "*.*")))
+filename = Path(filepath).stem
+current_time = datetime.now().strftime("%d-%m-%Y %H:%M")
+new_filepath = dir_path + filename + '_' + current_time + '_detected.mp4'
```
Comment (anatshk) [inline face_detection.py]: please use f-stirngs - f"{dir_path}/{filename}_{filename}'_detected.mp4'"

File: face_detection.py [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+
+v_cap = cv2.VideoCapture(filepath)
+frame_width = int(v_cap.get(3))
+frame_height = int(v_cap.get(4))
+v_len = int(v_cap.get(cv2.CAP_PROP_FRAME_COUNT))
+
+SMILE_MODEL_PATH = "state_dict_model_sensitivity.pt"
```
Comment (anatshk) [inline face_detection.py]: const

File: face_detection.py [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+        continue
+
+    for face in face_objects_list:
+        face: FaceObject
+        prediction = face.predictions_in_frames[i]
+        coordinates = face.coordinates_in_frames[i]
+        if prediction is not None:
+            pt_1 = int(coordinates[0]), int(coordinates[1])
+            pt_2 = int(coordinates[2]), int(coordinates[3])
+            color = (0, 255, 0) if prediction else (0, 0, 255)
```
Comment (anatshk) [inline face_detection.py]: the color tuplle should be a const - GREEN = (0, 255, 0)

File: face_detection.py [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+# if filename in df.file_name.values:
+#     labels_loc = df.loc[df.file_name == filename, 'face_1'].apply(eval).values[0]
+#     truth = np.zeros(labels_loc[-1])
+#     for i in range(0, len(labels_loc) - 1, 2):
+#         truth[labels_loc[i]+1 : labels_loc[i+1]] = [1] * (labels_loc[i+1] - labels_loc[i] - 1)
+#
+#     face_object: FaceObject = face_objects_list[0]
+#     accuracy = (truth == np.array(face_object.predictions_in_frames)).sum()
+#     accuracy = 100 * accuracy / v_len
+#     print('accuracy = {}'.format(accuracy))
```
Comment (anatshk) [inline face_detection.py]: delete commented-out code

File: face_detection.py [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+
+# open detected video
+if platform.system() == 'Darwin':       # macOS
+    subprocess.call(('open', new_filepath))
+elif platform.system() == 'Windows':    # Windows
+    os.startfile(new_filepath)
+else:                                   # linux variants
+    subprocess.call(('xdg-open', new_filepath))
```
Comment (anatshk) [inline face_detection.py]: if code is not tested and you know its usable - please delete

File: face_detection.py (hunk 9 above)
Comment (anatshk) [reply face_detection.py]: wrap all the code in one function + add CLI run (`if __name__ == "__main__"`)

File: smile_net.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+import torch.nn.functional as F
+
+
+class Net(nn.Module):
+  def __init__(self):
+    super(Net, self).__init__()
+    self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
+    self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
+    self.conv2_drop = nn.Dropout2d()
+    self.fc1 = nn.Linear(3380, 50)
```
Comment (anatshk) [inline smile_net.py]: what are these numbers?
if they are consts or a function of batch_size - please give them names

File: smile_recognition.py [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from tqdm import tqdm
+from sklearn.metrics import confusion_matrix
+from smile_net import Net
+
+torch.manual_seed(2)
+
+# def show(imgs):
```
Comment (anatshk) [inline smile_recognition.py]: commented out code - delete please

File: smile_recognition.py [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+
+class FaceDataset(Dataset):
+    def __init__(self, img_dir, transform=None):
+        self.img_dir = img_dir
+        self.img_paths = []
+        self.img_labels = []
+        self.transform = transform
+
+        positive_path = os.path.join(img_dir, "positives/positives7")
+        negative_path = os.path.join(img_dir, "negatives/negatives7")
```
Comment (anatshk) [inline smile_recognition.py]: paths here are hard-coded - maybe have inputs of `positives_path negatives_path`

File: smile_recognition.py [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+        for path in [positive_path, negative_path]:
+            for file in os.listdir(path):
+                if file.endswith(".jpg"):
+                    image_path = os.path.join(path, file)
+                    self.img_paths.append(image_path)
+                    if path is positive_path:
+                        self.img_labels.append(1)
+                    else:
+                        self.img_labels.append(0)
```
Comment (anatshk) [inline smile_recognition.py]: this should be a method of the class - "get img labels" - then run it on positives and negatives - not in loop, no reason for a loop of 2

File: smile_recognition.py [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+def generate_sampler(dataset, indices):
+    train_labels = []  # list of labels of any example in train set
+    class_counts = [0] * 2
+    num_samples = len(indices)
+    for idx in indices:
+        _, label = dataset[idx]
+        class_counts[label] += 1
+        train_labels.append(label)
+
+    class_weights = [num_samples / class_counts[i] for i in range(len(class_counts))]
```
Comment (anatshk) [inline smile_recognition.py]: `[num_samples / class_count for class_count in class_counts]`

File: smile_recognition.py [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+
+    # expected
+    if len(matrix) == 4:
+        return matrix
+
+    # else len = 1 which means that y_true = y_pred
+    elif y_true[0] == 0:
+        return [4, 0, 0, 0]
+    else:
+        return [0, 0, 0, 4]
```
Comment (anatshk) [inline smile_recognition.py]: else not needed - if + return --> else can be removed

File: smile_recognition.py [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+            false_positive += fp
+            false_negative += fn
+
+    accuracy = (100 * accuracy / total)
+    sensitivity = true_positive/(true_positive + false_negative)
+    specificity = true_negative/(true_negative + false_positive)
+    return accuracy, sensitivity, specificity
+
+
+img_dir = "../SMILEsmileD/SMILEs"
```
Comment (anatshk) [inline smile_recognition.py]: const

File: smile_recognition.py [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+
+transform = transforms.Compose([
```
Comment (anatshk) [inline smile_recognition.py:21]: please separate "run code" from classes and fiunctions

File: smile_recognition.py [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+])
+
+dataset = FaceDataset(img_dir, transform=transform)
+print("Numbers of images in dataset:", len(dataset))
+print("One example from dataset:", dataset[0])
+plt.imshow(dataset[0][0].permute(1, 2, 0))
+plt.show()
+
+train_size = int(0.8 * len(dataset))
+train_batch_size = 128
```
Comment (anatshk) [inline smile_recognition.py:41]: consts

File: smile_recognition.py [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+max_sensitivity = 0
+PATH = "state_dict_model_sensitivity.pt"
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+model.to(device)
+
+model.load_state_dict(torch.load(PATH, map_location=device))
+test_validation()
+
+for epoch in range(n_epochs):
```
Comment (anatshk) [inline smile_recognition.py]: function please - training?

File: smile_recognition.py [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+
+max_accuracy = 0
+max_sensitivity = 0
+PATH = "state_dict_model_sensitivity.pt"
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+model.to(device)
+
+model.load_state_dict(torch.load(PATH, map_location=device))
+test_validation()
```
Comment (anatshk) [inline smile_recognition.py]: delete? or move?
